File: SwapListPairwise.py
# -*- coding: utf-8 -*-
"""
Created on Tue Sep 29 09:35:55 2020

@author: funny
"""

"""
Given a singly linked list, write a function to swap elements pairwise.

Input : 1->2->3->4->5->6->NULL
Output : 2->1->4->3->6->5->NULL
"""

# Node class  
class Node: 
  
    # Constructor to initialize the node object 
    def __init__(self, data): 
        self.data = data 
        self.next = None
    def __str__(self):
        s = str(self.data) + "->"
        if(self.next):
            s += str(self.next)
        else:
            s += "None"
        return s
    
def swapPairwise(node):
    stack = []
    tmp = node
    while(tmp): #Add list to stack
        stack.append(tmp)
        tmp = tmp.next
    for i in range(0, len(stack), 2):
        #For each 2nd element in list swap positions
        if not (i+1 > len(stack)-1):
            stack.insert(i, (stack.pop(i+1)))
    for j in range(len(stack)-1): #Link together the newly swapped nodes
        stack[j].next = stack[j+1]
    stack[-1].next = None #Set the last node's next to None
    return stack[0]
    
node = Node(1)
node.next = Node(2)
node.next.next = Node(3)
node.next.next.next = Node(4)
node.next.next.next.next = Node(5)
#1->2->3->4->5->None

print("Before: " + str(node))
node = swapPairwise(node)
print("After: " + str(node))
#2->1->4->3->5->None

File: SwapListPairwise.java
/*
Given a singly linked list, write a function to swap elements pairwise.

Input : 1->2->3->4->5->6->NULL
Output : 2->1->4->3->6->5->NULL
*/

import java.util.ArrayList;

class SwapListPairwise{
   public static void main(String[] args){
      Node head = new Node(1);
      Node tmp = head;
      for(int i = 2; i <= 7; i++){
         tmp.next = new Node(i);
         tmp = tmp.next;
      }
      //1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7 -> null
      
      Node result = swapListPairwise(head);
      System.out.println(result);
      //2 -> 1 -> 4 -> 3 -> 6 -> 5 -> 7 -> null
   }
   
   /**
   * Swaps a given linked list pairwise.
   *
   * @param head The head of the linked list
   * @return The pairwise swapped linked list
   */
   static Node swapListPairwise(Node head){
      ArrayList<Node> arr = new ArrayList<Node>();
      Node tmp = head;
      while(tmp != null){
         arr.add(tmp);
         tmp = tmp.next;
      }
      //Pairwise swap elements
      for(int i = 0; i < arr.size()-1; i+=2){
         tmp = arr.get(i+1);
         arr.set(i+1, arr.get(i));
         arr.set(i, tmp);
      }
      //Set pointers
      for(int i = 0; i < arr.size()-1; i++){
         arr.get(i).next = arr.get(i+1);
      }
      arr.get(arr.size()-1).next = null;
      return arr.get(0);
   }
}

class Node{
   int data;
   Node next;
   
   public Node(int data){
      this.data = data;
   }
   
   public String toString(){
      return data + " -> " + next;
   }
}